core/src/main/java/org/apache/iceberg/io/RollingContentFileWriter.java

```diff
@@ -0,0 +1,152 @@
+
+  private static final int ROWS_DIVISOR = 1000;
+  private final PartitionKey partitionKey;
+  private final FileFormat format;
+  private final OutputFileFactory fileFactory;
+  private final FileIO io;
+  private final long targetFileSize;
+  private final ContentFileWriterFactory<ContentFileT, T> writerFactory;
+
+  private final WriterResult.Builder resultBuilder;
+
+  private EncryptedOutputFile currentFile = null;
+  private ContentFileWriter<ContentFileT, T> currentFileWriter = null;
+  private long currentRows = 0;
+
+  public RollingContentFileWriter(PartitionKey partitionKey, FileFormat format,
+                                  OutputFileFactory fileFactory, FileIO io,
+                                  long targetFileSize, ContentFileWriterFactory<ContentFileT, T> writerFactory) {
+    this.partitionKey = partitionKey;
+    this.format = format;
+    this.fileFactory = fileFactory;
+    this.io = io;
+    this.targetFileSize = targetFileSize;
+    this.writerFactory = writerFactory;
+
+    this.resultBuilder = WriterResult.builder();
+
+    openCurrent();
+  }
+
+  public CharSequence currentPath() {
+    return currentFile.encryptingOutputFile().location();
+  }
+
+  public long currentPos() {
+    return currentRows;
+  }
+
+  public void write(T record) throws IOException {
+    this.currentFileWriter.write(record);
+    this.currentRows++;
+
+    if (shouldRollToNewFile()) {
+      closeCurrent();
+      openCurrent();
+    }
+  }
+
+  public void abort() throws IOException {
+    close();
+
+    WriterResult result = resultBuilder.build();
+
+    Tasks.foreach(result.contentFiles())
+        .throwFailureWhenFinished()
+        .noRetry()
+        .run(file -> io.deleteFile(file.path().toString()));
+  }
+
+  public WriterResult complete() throws IOException {
+    close();
+
+    return resultBuilder.build();
+  }
+
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("partitionKey", partitionKey)
+        .add("format", format)
+        .toString();
+  }
+
+  private void openCurrent() {
+    if (partitionKey == null) {
+      // unpartitioned
+      currentFile = fileFactory.newOutputFile();
+    } else {
+      // partitioned
+      currentFile = fileFactory.newOutputFile(partitionKey);
+    }
+    currentFileWriter = writerFactory.createWriter(partitionKey, currentFile, format);
+    currentRows = 0;
+  }
+
+  private boolean shouldRollToNewFile() {
+    // TODO: ORC file now not support target file size before closed
+    return !format.equals(FileFormat.ORC) &&
+        currentRows % ROWS_DIVISOR == 0 && currentFileWriter.length() >= targetFileSize;
+  }
+
+  private void closeCurrent() throws IOException {
+    if (currentFileWriter != null) {
+      currentFileWriter.close();
+
+      ContentFileT contentFile = currentFileWriter.toContentFile();
+      Metrics metrics = currentFileWriter.metrics();
```

Comment:
This looks like the only use of `ContentFileWriter#metrics()`. I think it would be better to simplify this and just expose `rowCount()` instead of all metrics.

Reply:
Also, this class knows when a file has been rolled and keeps its own count of rows, so there is no need to get the number of rows passed to the current file writer. This should use `currentRows` instead. That way we can remove `metrics`.

Reply:
Yeah, the `currentRows` is enough, don't have to expose the `metrics` now.